### PR TITLE
[NFC] Refactor input tile loading and storing to LDS to be based on just views

### PR DESCRIPTION
**Pull request by manupak**
Currently there are a number of significant coupling between functions in gridwise to blockwise lowering that assumes a certain shapes and transforms. This kind of restricts usage of these for other gridwise kernels.

In this first refactor PR,

Im introducing a singular view creation function -- createGemmInputViewsFromGlobal --  that gives gridwise, blockwise and threadwise views of the global input tensors based on derivative tuning params.

Then refactor the rest code where input tile loading to just to rely on the those views generate further code.
In the process Im removing functions : 
* wrapMatrixForGlobalLoad (most of the view creation logic is lifted from here to createGemmInputViewsFromGlobal)
* globalVectorLayout - this is trying to re-generate threadwise view independently and not even used by createGlobalLoadLoop
* createGlobalLoadLoop - this seems identical to threadwise_read_into op.
* viewLoadBufferDK - I introduced this earlier to view loadedBuffer as DxK threadwise slice of the input tensor. However, it is just the inverted view of createGemmInputViewsFromGlobal.threadwiseView)

~~If these seems reasonable, I ll further refactor wrapLDSforStore (which is what I really want to be based on views) in the next PR~~

UPDATE : 

* I have done the further refactor to decouple wrapLDSStoreBuffer function from global tile loading assumption on this PR.
* Basically problem was wrapLDSStoreBuffer integrates knowledge from how the tile was loaded into regs and creates the views that it does not have control over. I've decoupled this now.

closes : https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/1029

### CODE REVIEW

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:243]: This is to be consistent with all types of views to end with ... x K x D

Comment (giuseros) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:243]: Does it mean we are transposing while loading?

Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:243]: no; it is a just a view of the load buffer.

Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:243]:
```c++
static void makeGemmInputViewsIter(TopDownTMBuilder &viewBuilder,
                                   StringRef dIterName, int64_t dPerThread,
                                   int64_t kPerThread,
                                   ArrayRef<unsigned> outDims,
                                   bool isKContigousDim) {
  if (isKContigousDim) {
    viewBuilder.merge({dIterName, "k_iter"}, outDims, "iter",
                      {dPerThread, kPerThread});
  } else {
    viewBuilder.merge({"k_iter", dIterName}, outDims, "iter",
                      {kPerThread, dPerThread});
  }
}
```
i.e. how iter is flattened will differ based on what is the contiguous dim

File: mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h]: Can we get some comments here on what these views actually are?

**Review by krzysz00 — COMMENTED** [review]
I'm broadly happy here aside from some documentation.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:257]: Ive removed the vectorDim now

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: We create the re-shuffled regs views first throughout (not just threadwise).
Then we transform the storeBuffer with inverted threadwise view that will produce K x D view of whats inside the regs.
This makes this composable with the K x D view of the loadBuffer.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:670]: Now wrapping here will produce K x D to flatLDS byte buffer because (tid, iter) to K x D is defined by the tile loading process , not here.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: Here we compose : (tid, iter) --> K x D (defined by tile loading process) and K x D --> flatLDSbuffer

**Review by manupak — COMMENTED** [review]
I've done somewhat significant refactor here.. and put down some comments to follow the PR. 
Since it was all in my head, I m happy to anwser questions / improve docs.

The main gist is I've decoupled view creations to always be based on the operator space which is K x D (basically logical tile dims). Then construct & compose views to reg/lds buffers as needed. This basically allows to reuse much of the components with the operations that can produce K x D views.

File: mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h
Comment (manupak) [reply mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h]: I've added for these and others here.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:670]: or it could be defined by a prior gemm for e.g.

File: mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h]: This needs more clarity.

Tile of what? What'll might use these outputs? What are K and D? Are they the same for all of these?

File: mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h]: This struct name is _way_ too generic. `LoadedTileViews`? `GridToMatrixMaps`?

File: mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h]: Somewhere, either in this file or elsewhere, define what this function is and how you're meant to use it.

File: mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h]: I do not understand this comment

File: mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h]: Returning a `StringRef` is extremely suspicious on memory safety grounds.

File: mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h]: These dim part info things should probably be taken by reference, they're big

File: mlir/lib/Dialect/Rock/utility/loweringUtils.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/utility/loweringUtils.cpp]: Take these as out parameters of `SmallVectorImpl`

File: mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h
Comment (krzysz00) [reply mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h]: (It'll probably work, but I'd be careful here if anyone is constructing derived strings from this value. If nothing else, warn about it)

File: mlir/lib/Dialect/Rock/utility/loweringUtils.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/utility/loweringUtils.cpp]: I'm a bit suspicious of this

This looks like uniformly assuming that the iter dimension moves faster than the tid dimension, which doesn't feel right.

File: mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h]: I'm not liking these.

It doesn't seem like you can quite accomplish what you're going for by just returning a `rock::TransformAttr` and some result sizes, but ... maybe? 

Especially if the transform map builders grew an `insertTransform(TransformAttr, ArrayRef<int64_t>)`.

File: mlir/lib/Dialect/Rock/utility/loweringUtils.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/utility/loweringUtils.cpp]: I'm very suspicious of the fact that you need this.

**Review by krzysz00 — CHANGES_REQUESTED** [review]
I'm sticking a block on this because I don't understand it and it's just ... really suspicious. 

It feels both overly general and like it's making a lot of strong assumptions about clients at the same time.

I'm also calling for improved documentation, but I'm not sure it'd be good to write it until we're happy with the design.

File: mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h
Comment (manupak) [reply mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h]: These are an artifact that we cant create a graph of transforms rather than a chain of transform. 

I can flat out create two independent view creators here : 
1) loaded tile from global
2) reshuffled tile from global for lds store

Then remove these utils. 
I did this to factor the common thing between the above two.

File: mlir/lib/Dialect/Rock/utility/loweringUtils.cpp
Comment (manupak) [reply mlir/lib/Dialect/Rock/utility/loweringUtils.cpp]: it works between :

1) loaded tile from global
2) reshuffled tile from global

and yea it wont work for others -- not as general that bit.
I could just simply make two independent view creators.

File: mlir/lib/Dialect/Rock/utility/loweringUtils.cpp
Comment (manupak) [reply mlir/lib/Dialect/Rock/utility/loweringUtils.cpp]: The fact this was needed to factor out create similiar views for 1) loaded tile from global and 2) reshuffled tile from global where the difference is only how iter maps to indices in regs.

**Comment by manupak** [comment]
I think broadly the complaints are about the attempt in trying to factor out two view generators which might as well be independent of each other though they share partial coord transform of bid and tid but not iter.  I might remove all that utils and create two seperate functions : 
1) createGemmInputTileViewsFromGlobal
2) createGemmInputTileViewsToLds

please confirm if that would clear things up for you.

**Comment by krzysz00** [comment]
I'd like to understand what it is you meant by "input tile" in those function names before I say anything concrete.

**Comment by manupak** [comment]
Input tile is the tile of the input passed to the gemm. It is two dimensional but 3 different sizes when you look at it from gridwise, blockwise, and threadwise level.

Gridwise: this is the largest 2d tile all blocks collectively hold in the GPU
Blockwise: this is the tile all threads in a block collectively hold in the GPU
Threadwise: this is the tile each thread hold

After transformations, we keep track of those views.

**Comment by krzysz00** [comment]
Ok, so, because now that I've slept on things I think it's important to get some more names in here because I'm having trouble what's going on here.

For simplicity, I'm leaving the `G` dimension out of all of this.

First, some variable names and their ranges (they may not match the code or the transform dims, but)
- `kg : [0, K)` a coordinate in the K dimension of the global tensor
- `dg : [0, M/N)` a coordinate in the D (M/N) dimension of the global tensor
- `kbl : [0, KPerBlock)` a coordinate in the K dimension of the tile loaded by a workgroup
- `dbl : [0, DPerBlock)` same for the D dimension
- `kObs : [0, KPacksPerBlock)` a coordinate in the kOuter dimension of the LDS matrix
- `dbs : [0, DPerBlock)` same for the D dimension
- `kPbs : [0, KPack)` same for kPack
- `krl : [0, KPerThread)` a coordinate in the registers loaded from global memory, in the K dimension.
- `drl: [0, DPerThread)` similarly
- `kOrs : `[0, KOuterPerThread)`: a coordinate in the registers as they're being stored to LDS, KOuter dimension
- `drs: [0, DPerThread)`: similarly
- `kPrs : [0, kPackPerThread)`: similarly
- `...bid`: The coordinates representing the block ID
- `tid : [0, blockSize)`: ID of a thread within a workgroup
- `iter : [0, kPerThread * dPerThread = R)` the number of values each thread posseses

We have, or might want  have, the following functions
- `toGlobalMatrixCoords(... bid, tid, iter) -> (kg, dg)` (... to actual global memory)
- `loadRegsAsThreadTile(iter) -> (krl, drl)` We don't currently have this. In your case you'll probably want `loadRegsAsThreadTile(iter) -> (krl, drl)` and `accelRegsAsThreadTile(iter) -> (krl, drl)` because they could be distinct
  - Or, if I've correctly understood the generalization you're going for, there's `loadRegsAsThreadTile(tid, iter) -> (krl, drl)` and `accelRegsAsThreadTile(tid, iter) -> (krl, drl)`
- We have `threadTileToPackedThreadTile(krl, drl) -> (kOrs, drs, kPrs)`, which handles the packing and transposes
- We have `storeRegAsPackedThreadTile(kOrs, drs, kPrs) -> (iter)`
- We have `toLDSMatrixCoords(tid, iter) -> (kObs, dbs, kPbs)`

I think some of your changes come down to making those `RegsAsTile` functions go the other way: that is, we'll define `threadTileToLoadRegs(krl, drl) -> (iter)` and that sort of thing.

Have I missed some functions you need?

**Comment by manupak** [comment]
Apart from naming, we are mostly talking about the same thing except : 

> We have threadTileToPackedThreadTile(krl, drl) -> (kOrs, drs, kPrs), which handles the packing and transposes
> We have storeRegAsPackedThreadTile(kOrs, drs, kPrs) -> (iter)
> We have toLDSMatrixCoords(tid, iter) -> (kObs, dbs, kPbs)

One of the qualities I want to preserve at each stage of lowering, I'd like to see that as a tile that is going into the gemm 'logically'. 
e.g. instead of threadTileToPackedThreadTile (krl, drl) -> (kOrs, drs, kPrs): 

we do : 
iter -->  (kOrs, drs, kPrs) --> (krl, drl)  
Lets call it packedRegsAsThreadTile

This view tells me what what the registers holds after the packing as well as 'how' it is being held.
Moreover, it allows me to do the packing as follows :
i.e. 
loadRegsAsThreadTile[loadRawBuffer] --> packedRegsAsThreadTile[storeRawBuffer]

Now because I have that view -- packedRegsAsThreadTile -- I can compose this further.

Extending that quality if every macro op signoff using that logical shape views -- we can keep composing ops.

**Comment by manupak** [comment]
So packedRegAsThreadTile is no different in structure to loadRegsAsThreadTile or accelRegsAsThreadTile

**Comment by manupak** [comment]
On top of that I want 

XAsThreadTile
XAsBlockTile
XAsGridTile

to be generated simultaneously so we have them at every point of lowering without one going further than the others...

**Comment by manupak** [comment]
End of the day, following your terminology, I'd like to have : 
1) loadRegsAsThreadTile, loadRegsAsBlockTile, loadRegsAsGridTile
2) packedRegsAsThreadTile, packedRegsAsBlockTile, packedRegsAsGridTile
3) accelRegsAsThreadTile, accelRegsAsBlockTile, accelRegsAsGridTile

Since I wanted each of them to be simultaneously generated : 

1) loadRegsAsTileViews(...) -> GPUViews
2) packedRegsAsTileViews(...) -> GPUViews
3) accelRegsAsTileViews(...) --> GPUViews

**Comment by manupak** [comment]
Now Ive done : 
* GPUViews --> RegsAsMatrixSubTiles
* created getLoadRegsAsTileViews
* created getPackedRegsAsTileViews

and removed the rest of the utils as discussed.
PTAL @krzysz00;

**Review by krzysz00 — APPROVED** [review]
This all looks pretty comprehensible ... approving, but I'd like an assembly diff check (and a quick perf check) before landing this just to be safe.

**Comment by manupak** [comment]
Yes -- agreed.
The gemms look flat but convs have regressions.
The regressions are for conv nchw x [i8, f16] cases which shouldn't be there.
 Im looking into it what causes that.

[PR1164_perf.zip](https://github.com/ROCmSoftwarePlatform/rocMLIR/files/12239718/PR1164_perf.zip)

**Comment by manupak** [comment]
[PR1164_conv_perf16.zip](https://github.com/ROCmSoftwarePlatform/rocMLIR/files/12247895/PR1164_conv_perf16.zip)

Ok conv perfs are more or less flat now.. It turned out to be that vector length inference is not working for all the cases for global loads when its looked at after the input is viewed as gridSubTile. I just added a forceVecLen option for threadwise_read_into workaround the issue and filed a bug report : 
https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/1049

This made the perf flat as expected.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:213]: In my analysis (mostly IR diffs), I have found and fixed this bug currently; we use m_blocks as n_blocks.

**Comment by krzysz00** [comment]
Ok, wait, so does the vectorization analysis still need to be tweaked or?

**Comment by manupak** [comment]
So the current vectorization analysis "works" on the view/memref present to the op (though our perf tests does only checks tranposes realistically, aren't they?). 

However, if I do further views on it (i.e. gridSubTile) it kind of breaks.

So there are only two explanations : 
1) either my views are wrong
2) or vectorization analysis doesn't work in some cases

If you are suspecting 1) I fail to see how that would produce the correct result with forced vectorization.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:385]: Also I'd like to point out that vectorLen is already forced in the current code where there isn't vec len inference done after untransform. So this PR doesn't change this aspect.

Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:385]: I can reproduce the same issue with the following changes to develop as well...

```c++
--- a/mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
+++ b/mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
@@ -359,20 +359,26 @@ static ArrayAttr ldsVectorLayout(OpBuilder &b, Location loc,
 static TransformingForOp
 createGlobalLoadLoop(PatternRewriter &b, Location loc, GpuAllocOp loadBuffer,
                      Value wrappedMatrix, ArrayAttr vectorMap,
-                     int64_t dataPerThread, int64_t vectorLen, Value bid,
+                     int64_t dataPerThread, int64_t vectorLenExt, Value bid,
                      Value tid, layout::GridCoordinates coords,
                      bool forceUnroll) {
   Value tensor;
   ArrayAttr matrixToTensor;
   std::tie(tensor, matrixToTensor) = untransform(b, wrappedMatrix);
+  ArrayRef<int64_t> tensorSize = tensor.getType().cast<ShapedType>().getShape();
+
+  ArrayRef<int64_t> wrappedMatrixShape =
+      wrappedMatrix.getType().cast<ShapedType>().getShape();
+  Type elemType = wrappedMatrix.getType().cast<MemRefType>().getElementType();
+  int64_t vecLen = getMaxVectorizationForDatatype(matrixToTensor, /*dim=*/5,
+                                          wrappedMatrixShape[5], tensorSize, elemType);

   // Optimize the transform chain.
-  ArrayRef<int64_t> tensorSize = tensor.getType().cast<ShapedType>().getShape();
   matrixToTensor = collapseContiguousMerges(matrixToTensor, tensorSize);

   Type elementType =
       wrappedMatrix.getType().cast<MemRefType>().getElementType();
-  Type loadType = vectorTypeOrSelf(elementType, vectorLen);
+  Type loadType = vectorTypeOrSelf(elementType, vecLen);
   Value zero = b.createOrFold<arith::ConstantIndexOp>(loc, 0);
```

Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:385]: rocmlir-gen --operation conv2d -t i8 --arch gfx908:sramecc+:xnack- --num_cu 120 --fil_layout kcyx --in_layout nchw --out_layout nkhw --batchsize 256 --in_channels 64 --in_h 56 --in_w 56 --out_channels 256 --fil_h 1 --fil_w 1 --dilation_h 1 --dilation_w 1 --conv_stride_h 1 --conv_stride_w 1 --padding_h 0 --padding_w 0 --kernel-repeats 5 --perf_config=32,128,32,32,32,1,1,1 | rocmlir-driver -c --mlir-print-ir-before=rock-blockwise-gemm-to-threadwise &> dump_br.log

**Comment by manupak** [comment]
Ok, I've reverted back forceVec changes and included the commit from https://github.com/ROCmSoftwarePlatform/rocMLIR/pull/1179 to check the CI.

Perfs are flat now as well...
[PR1164_conv_perf17.zip](https://github.com/ROCmSoftwarePlatform/rocMLIR/files/12261123/PR1164_conv_perf17.zip)

**Comment by manupak** [comment]
I've taken this in on the basis it was approved code (at the point of approval) and perfs are flat.

**Pull request merged**
